File: projects/php-sdk/src/Controllers/EventsController.php
<?php

namespace PHP_SDK\Controllers;

use Fingerprint\ServerAPI\Api\FingerprintApi;
use Fingerprint\ServerAPI\ApiException;
use Fingerprint\ServerAPI\Configuration;
use Fingerprint\ServerAPI\Model\EventsUpdateRequest;
use GuzzleHttp\Client;
use GuzzleHttp\Psr7\Response;
use GuzzleHttp\Psr7\ServerRequest;
use PHP_SDK\Models\MusicianResponse;
use Psr\Http\Message\MessageInterface;

class EventsController
{
    public function searchEvents(ServerRequest $request, Response $response): MessageInterface {
        $queryParams = $request->getQueryParams();
        $apiKey = $queryParams['apiKey'] ?? '';
        $region = $queryParams['region'] ?? '';
        $limit = $queryParams['limit'] ?? null;

        $paginationKey = $queryParams['paginationKey'] ?? null;
        $visitorId = $queryParams['visitorId'] ?? null;
        $bot = $queryParams['bot'] ?? null;
        $ipAddress = $queryParams['ipAddress'] ?? null;
        $linkedId = $queryParams['linkedId'] ?? null;
        $start = isset($queryParams['start']) ? (int) $queryParams['start'] : null;
        $end = isset($queryParams['end']) ? (int) $queryParams['end'] : null;
        $reverse = isset($queryParams['reverse']) ? filter_var($queryParams['reverse'], FILTER_VALIDATE_BOOLEAN) : null;
        $suspect = isset($queryParams['suspect']) ? filter_var($queryParams['suspect'], FILTER_VALIDATE_BOOLEAN) : null;
        $vpn = isset($queryParams['vpn']) ? filter_var($queryParams['vpn'], FILTER_VALIDATE_BOOLEAN) : null;
        $virtualMachine = isset($queryParams['virtualMachine']) ? filter_var($queryParams['virtualMachine'], FILTER_VALIDATE_BOOLEAN) : null;
        $tampering = isset($queryParams['tampering']) ? filter_var($queryParams['tampering'], FILTER_VALIDATE_BOOLEAN) : null;
        $antiDetectBrowser = isset($queryParams['antiDetectBrowser']) ? filter_var($queryParams['antiDetectBrowser'], FILTER_VALIDATE_BOOLEAN) : null;
        $incognito = isset($queryParams['incognito']) ? filter_var($queryParams['incognito'], FILTER_VALIDATE_BOOLEAN) : null;
        $privacySettings = isset($queryParams['privacySettings']) ? filter_var($queryParams['privacySettings'], FILTER_VALIDATE_BOOLEAN) : null;
        $jailbroken = isset($queryParams['jailbroken']) ? filter_var($queryParams['jailbroken'], FILTER_VALIDATE_BOOLEAN) : null;
        $frida = isset($queryParams['frida']) ? filter_var($queryParams['frida'], FILTER_VALIDATE_BOOLEAN) : null;
        $factoryReset = isset($queryParams['factoryReset']) ? filter_var($queryParams['factoryReset'], FILTER_VALIDATE_BOOLEAN) : null;
        $clonedApp = isset($queryParams['clonedApp']) ? filter_var($queryParams['clonedApp'], FILTER_VALIDATE_BOOLEAN) : null;
        $emulator = isset($queryParams['emulator']) ? filter_var($queryParams['emulator'], FILTER_VALIDATE_BOOLEAN) : null;
        $rootApps = isset($queryParams['rootApps']) ? filter_var($queryParams['rootApps'], FILTER_VALIDATE_BOOLEAN) : null;
        $minSuspectScore = isset($queryParams['minSuspectScore']) ? (float) $queryParams['minSuspectScore'] : null;
        $ipBlocklist = isset($queryParams['ipBlocklist']) ? filter_var($queryParams['ipBlocklist'], FILTER_VALIDATE_BOOLEAN) : null;
        $datacenter = isset($queryParams['datacenter']) ? filter_var($queryParams['datacenter'], FILTER_VALIDATE_BOOLEAN) : null;

        $config = Configuration::getDefaultConfiguration($apiKey, $region);
        $client = new FingerprintApi(
            new Client(),
            $config
        );

        try {
            list($model, $apiResponse) = $client->searchEvents(
                $limit,
                pagination_key: $paginationKey,
                visitor_id: $visitorId,
                bot: $bot,
                ip_address: $ipAddress,
                linked_id: $linkedId,
                start: $start,
                end: $end,
                reverse: $reverse,
                suspect: $suspect,
                vpn: $vpn,
                virtual_machine: $virtualMachine,
                tampering: $tampering,
                anti_detect_browser: $antiDetectBrowser,
                incognito: $incognito,
                privacy_settings: $privacySettings,
                jailbroken: $jailbroken,
                frida: $frida,
                factory_reset: $factoryReset,
                cloned_app: $clonedApp,
                emulator: $emulator,
                root_apps: $rootApps,
                min_suspect_score: $minSuspectScore,
                ip_blocklist: $ipBlocklist,
                datacenter: $datacenter
            );

            $result = new MusicianResponse($apiResponse->getStatusCode(), $apiResponse, $model);
        } catch (ApiException $e) {
            $result = MusicianResponse::BuildForApiException($e);
        } catch (\Throwable $e) {
            $result = new MusicianResponse(500, $e->getMessage(), $e->getMessage());
        }

        $response->getBody()->write(json_encode($result));

        return $response->withHeader('Content-Type', 'application/json');
    }

    public function getEvents(ServerRequest $request, Response $response): MessageInterface
    {
        $queryParams = $request->getQueryParams();
        $apiKey = $queryParams['apiKey'] ?? '';
        $region = $queryParams['region'] ?? '';
        $requestId = $queryParams['requestId'] ?? '';

        $config = Configuration::getDefaultConfiguration($apiKey, $region);
        $client = new FingerprintApi(
            new Client(),
            $config
        );

        try {
            list($model, $api_response) = $client->getEvent($requestId);
            $result = new MusicianResponse($api_response->getStatusCode(), $api_response, $model);
        } catch (ApiException $e) {
            $result = MusicianResponse::BuildForApiException($e);
        } catch (\Throwable $e) {
            $result = new MusicianResponse(500, $e->getMessage(), $e->getMessage());
        }


        $response->getBody()->write(json_encode($result));

        return $response->withHeader('Content-Type', 'application/json');
    }

    public function updateEvent(ServerRequest $request, Response $response): MessageInterface
    {
        $queryParams = $request->getQueryParams();
        $apiKey = $queryParams['apiKey'] ?? '';
        $region = $queryParams['region'] ?? '';
        $requestId = $queryParams['requestId'] ?? '';

        $linkedId = $queryParams['linkedId'] ?? '';
        $suspect = $queryParams['suspect'] ?? null;
        $tag = $queryParams['tag'] ?? null;

        $config = Configuration::getDefaultConfiguration($apiKey, $region);
        $client = new FingerprintApi(
            new Client(),
            $config
        );

        $body = new EventsUpdateRequest();
        if ($linkedId) {
            $body->setLinkedId($linkedId);
        }

        if ($suspect) {
            $body->setSuspect(filter_var($suspect, FILTER_VALIDATE_BOOLEAN, FILTER_NULL_ON_FAILURE));
        }

        if ($tag) {
            $parsedTag = json_decode($tag, true);
            $body->setTag($parsedTag);
        }

        try {
            list($model, $api_response) = $client->updateEvent($body, $requestId);
            $result = new MusicianResponse($api_response->getStatusCode(), $api_response, $model);
        } catch (ApiException $e) {
            $result = MusicianResponse::BuildForApiException($e);
        } catch (\Throwable $e) {
            $result = new MusicianResponse(500, $e->getMessage(), $e->getMessage());
        }

        $response->getBody()->write(json_encode($result));

        return $response->withHeader('Content-Type', 'application/json');
    }
}

File: projects/dotnet-sdk/Controllers/V4/EventsController.cs
using System.Net;
using System.Text.Json;
using Microsoft.AspNetCore.Mvc;
using Fingerprint.ServerSdk.Api;
using Fingerprint.ServerSdk.Client;
using Fingerprint.ServerSdk.Model;
using dotnet_sdk.Models;

namespace dotnet_sdk.Controllers.V4;

[ApiController]
[Route("v4")]
public class EventsController(FingerprintV4Factory factory) : ControllerBase
{
    [HttpGet("getEvent")]
    public async Task<IActionResult> GetEvent(
        [FromQuery(Name = "api_key")] string? apiKey,
        [FromQuery(Name = "region")] string? region,
        [FromQuery(Name = "event_id")] string? eventId,
        [FromQuery(Name = "ruleset_id")] string? rulesetId)
    {
        try
        {
            var api = factory.CreateApi(apiKey, region);

            var rulesetIdOption = rulesetId != null
                ? new Option<string>(rulesetId)
                : default;

            var apiResponse = await api.GetEventAsync(eventId ?? "", rulesetIdOption);

            return apiResponse.TryOk(out var data)
                ? Ok(new MusicianResponse<Event>(HttpStatusCode.OK, apiResponse.RawContent, data))
                : Ok(new MusicianResponse<object>(apiResponse.StatusCode, apiResponse.RawContent, Utils.ParseRawContent(apiResponse.RawContent)));
        }
        catch (Exception e)
        {
            return Ok(Utils.ProcessException(e));
        }
    }

    [HttpGet("searchEvents")]
    public async Task<IActionResult> SearchEvents(
        [FromQuery(Name = "api_key")] string? apiKey,
        [FromQuery(Name = "region")] string? region,
        [FromQuery(Name = "limit")] int? limit,
        [FromQuery(Name = "pagination_key")] string? paginationKey,
        [FromQuery(Name = "visitor_id")] string? visitorId,
        [FromQuery(Name = "bot")] string? bot,
        [FromQuery(Name = "ip_address")] string? ipAddress,
        [FromQuery(Name = "linked_id")] string? linkedId,
        [FromQuery(Name = "start")] long? start,
        [FromQuery(Name = "end")] long? end,
        [FromQuery(Name = "reverse")] bool? reverse,
        [FromQuery(Name = "suspect")] bool? suspect,
        [FromQuery(Name = "vpn")] bool? vpn,
        [FromQuery(Name = "virtual_machine")] bool? virtualMachine,
        [FromQuery(Name = "tampering")] bool? tampering,
        [FromQuery(Name = "anti_detect_browser")] bool? antiDetectBrowser,
        [FromQuery(Name = "incognito")] bool? incognito,
        [FromQuery(Name = "privacy_settings")] bool? privacySettings,
        [FromQuery(Name = "jailbroken")] bool? jailbroken,
        [FromQuery(Name = "frida")] bool? frida,
        [FromQuery(Name = "factory_reset")] bool? factoryReset,
        [FromQuery(Name = "cloned_app")] bool? clonedApp,
        [FromQuery(Name = "emulator")] bool? emulator,
        [FromQuery(Name = "root_apps")] bool? rootApps,
        [FromQuery(Name = "min_suspect_score")] float? minSuspectScore,
        [FromQuery(Name = "ip_blocklist")] bool? ipBlocklist,
        [FromQuery(Name = "datacenter")] bool? datacenter,
        [FromQuery(Name = "developer_tools")] bool? developerTools,
        [FromQuery(Name = "location_spoofing")] bool? locationSpoofing,
        [FromQuery(Name = "mitm_attack")] bool? mitmAttack,
        [FromQuery(Name = "proxy")] bool? proxy,
        [FromQuery(Name = "sdk_version")] string? sdkVersion,
        [FromQuery(Name = "sdk_platform")] string? sdkPlatform,
        [FromQuery(Name = "environment")] List<string>? environment,
        [FromQuery(Name = "proximity_id")] string? proximityId,
        [FromQuery(Name = "vpn_confidence")] string? vpnConfidence,
        [FromQuery(Name = "asn")] string? asn,
        [FromQuery(Name = "url")] string? url,
        [FromQuery(Name = "origin")] string? origin,
        [FromQuery(Name = "bundle_id")] string? bundleId,
        [FromQuery(Name = "tor_node")] bool? torNode,
        [FromQuery(Name = "package_name")] string? packageName
    )
    {
        try
        {
            var api = factory.CreateApi(apiKey, region);

            var request = new SearchEventsRequest { Limit = limit ?? 10 };

            if (paginationKey != null) request = request.WithPaginationKey(paginationKey);
            if (visitorId != null) request = request.WithVisitorId(visitorId);
            if (linkedId != null) request = request.WithLinkedId(linkedId);
            if (ipAddress != null) request = request.WithIpAddress(ipAddress);
            if (start.HasValue) request = request.WithStart(start.Value);
            if (end.HasValue) request = request.WithEnd(end.Value);
            if (reverse.HasValue) request = request.WithReverse(reverse.Value);
            if (suspect.HasValue) request = request.WithSuspect(suspect.Value);
            if (vpn.HasValue) request = request.WithVpn(vpn.Value);
            if (virtualMachine.HasValue) request = request.WithVirtualMachine(virtualMachine.Value);
            if (tampering.HasValue) request = request.WithTampering(tampering.Value);
            if (antiDetectBrowser.HasValue) request = request.WithAntiDetectBrowser(antiDetectBrowser.Value);
            if (incognito.HasValue) request = request.WithIncognito(incognito.Value);
            if (privacySettings.HasValue) request = request.WithPrivacySettings(privacySettings.Value);
            if (jailbroken.HasValue) request = request.WithJailbroken(jailbroken.Value);
            if (frida.HasValue) request = request.WithFrida(frida.Value);
            if (factoryReset.HasValue) request = request.WithFactoryReset(factoryReset.Value);
            if (clonedApp.HasValue) request = request.WithClonedApp(clonedApp.Value);
            if (emulator.HasValue) request = request.WithEmulator(emulator.Value);
            if (rootApps.HasValue) request = request.WithRootApps(rootApps.Value);
            if (minSuspectScore.HasValue) request = request.WithMinSuspectScore(minSuspectScore.Value);
            if (developerTools.HasValue) request = request.WithDeveloperTools(developerTools.Value);
            if (locationSpoofing.HasValue) request = request.WithLocationSpoofing(locationSpoofing.Value);
            if (mitmAttack.HasValue) request = request.WithMitmAttack(mitmAttack.Value);
            if (proxy.HasValue) request = request.WithProxy(proxy.Value);
            if (bot != null)
            {
                // To prevent internal error, just response like Server API error
                try { request = request.WithBot(SearchEventsBotValueConverter.FromString(bot)); }
                catch { throw new ApiException("Invalid bot filter", HttpStatusCode.BadRequest, "{\"error\":{\"code\":\"request_cannot_be_parsed\",\"message\":\"invalid bot type\"}}"); }
            }
            if (sdkVersion != null) request = request.WithSdkVersion(sdkVersion);
            if (sdkPlatform != null)
            {
                // To prevent internal error, just response like Server API error
                try { request = request.WithSdkPlatform(SearchEventsSdkPlatformValueConverter.FromString(sdkPlatform)); }
                catch { throw new ApiException("Invalid sdkPlatform", HttpStatusCode.BadRequest, "{\"error\":{\"code\":\"request_cannot_be_parsed\",\"message\":\"invalid sdk platform\"}}"); }
            }
            if (vpnConfidence != null)
            {
                // To prevent internal error, just response like Server API error
                try { request = request.WithVpnConfidence(SearchEventsVpnConfidenceValueConverter.FromString(vpnConfidence)); }
                catch { throw new ApiException("Invalid vpnConfidence", HttpStatusCode.BadRequest, "{\"error\":{\"code\":\"request_cannot_be_parsed\",\"message\":\"invalid vpn confidence\"}}"); }
            }
            if (environment is { Count: > 0 }) request = request.WithEnvironment(environment);
            if (proximityId != null) request = request.WithProximityId(proximityId);
            if (asn != null) request = request.WithAsn(asn);
            if (url != null) request = request.WithUrl(url);
            if (origin != null) request = request.WithOrigin(origin);
            if (bundleId != null) request = request.WithBundleId(bundleId);
            if (torNode.HasValue) request = request.WithTorNode(torNode.Value);
            if (packageName != null) request = request.WithPackageName(packageName);

            var apiResponse = await api.SearchEventsAsync(request);

            return apiResponse.TryOk(out var data)
                ? Ok(new MusicianResponse<EventSearch>(HttpStatusCode.OK, apiResponse.RawContent, data))
                : Ok(new MusicianResponse<object>(apiResponse.StatusCode, apiResponse.RawContent, Utils.ParseRawContent(apiResponse.RawContent)));
        }
        catch (Exception e)
        {
            return Ok(Utils.ProcessException(e));
        }
    }

    [HttpGet("updateEvent")]
    public async Task<IActionResult> UpdateEvent(
        [FromQuery(Name = "api_key")] string? apiKey,
        [FromQuery(Name = "region")] string? region,
        [FromQuery(Name = "event_id")] string? eventId,
        [FromQuery(Name = "linked_id")] string? linkedId,
        [FromQuery(Name = "tags")] string? tags,
        [FromQuery(Name = "suspect")] bool? suspect)
    {
        try
        {
            var api = factory.CreateApi(apiKey, region);

            var updateRequest = new EventUpdate();
            if (linkedId != null) updateRequest.LinkedId = linkedId;
            if (suspect.HasValue) updateRequest.Suspect = suspect;
            if (tags != null) updateRequest.Tags = JsonSerializer.Deserialize<Dictionary<string, object>>(tags)!;

            var apiResponse = await api.UpdateEventAsync(eventId ?? "", updateRequest);

            return Ok(new MusicianResponse<object>(apiResponse.StatusCode, apiResponse.RawContent ?? "", Utils.ParseRawContent(apiResponse.RawContent)));
        }
        catch (Exception e)
        {
            return Ok(Utils.ProcessException(e));
        }
    }
}
